### PR TITLE
[ML] adding new entropy custom feature processor for data frame analytics

### PR DESCRIPTION
**Pull request by benwtrent**
This adds a new Shannon entropy encoder for data frame analytics and inference.

Shannon entropy is a simple encoding for gathering the information provided inside a single string. 

Example:

"aaaaa" Provides no information
"babababababa" Provides 1 bit of information
"thisisarandomstring" is 3.3 bits of information

### CODE REVIEW

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/inference/preprocessing/Entropy.java [hunk 1]
```diff
@@ -0,0 +1,219 @@
...
+        fields.put(featureName, entropy);
+    }
+
+    double entropyCalc(String stringValue) {
+        final double len = stringValue.length();
+        // This provides a nice optimization for throughput.
+        // The entire basic latin code block fits is within 0x0000..0x007F.
+        // So, this provides a good optimization for those types of characters
+        // We still have a sparse hashmap for the rest.
+        int[] vals = new int[OPTIMIZATION_SIZE];
```
Comment (benwtrent) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/inference/preprocessing/Entropy.java:128]: One might suspect that this optimization is not worth it, but at the small expense of 127*4 bytes, we get almost double the speed.

```
# Benchmark on number of microseconds per op.
# This benchmark was the current hybrid approproach vs just a pre-allocated HashMap
# This is the average speed per ops for going through many strings of various sizes and languages
Benchmark                     Mode  Cnt    Score   Error  Units
EntropyBenchmark.hybrid       avgt   15  242.901 ± 4.729  us/op
EntropyBenchmark.unoptimized  avgt   15  405.232 ± 3.833  us/op
```

**Review by benwtrent — COMMENTED** [review]
I am opting for docs in a separate PR. We need a whole new page to cover our processors as we add them.

//CC @lcawl @szabosteve

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/inference/preprocessing/Entropy.java [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+        Map<Integer, Integer> biggerCounts = new HashMap<>();
+        for (PrimitiveIterator.OfInt it = stringValue.codePoints().iterator(); it.hasNext(); ) {
```
Comment (davidkyle) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/inference/preprocessing/Entropy.java:130]:
```suggestion
        stringValue.codePoints().foreach( cp -> {
```

**Review by davidkyle — APPROVED** [review]
LGTM

**Pull request closed without merging**